### PR TITLE
tools/rados: add support for binary object names in the rados CLI

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by ivancich — APPROVED** [review]
LGTM. Thank you. I did not suspect this change would be that involved.

File: src/tools/rados/rados.cc [hunk 1]
```diff
@@ -3312,27 +3376,26 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
   } else if (strcmp(nargs[0], "set-alloc-hint") == 0) {
-    if (!pool_name || nargs.size() < 4) {
+    if (!pool_name || nargs.size() < (obj_name ? 3 : 4)) {
       usage(cerr);
       return 1;
     }
     string err;
-    string oid(nargs[1]);
-    uint64_t expected_object_size = strict_strtoll(nargs[2], 10, &err);
+    uint64_t expected_object_size = strict_strtoll(nargs[obj_name ? 1 : 2], 10, &err);
```
Comment (tchaikov) [inline src/tools/rados/rados.cc:3384]: how about:
```c++
size_t arg_index = 1;
string oid;
if (obj_name) {
  oid = *obj_name;
} else {
  oid = nargs[arg_index++];
}
uint64_t expected_object_size = strict_strtoll(nargs[arg_index++], 10, &err);
// ...
```

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -3312,27 +3376,26 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
     }
-    uint64_t expected_write_size = strict_strtoll(nargs[3], 10, &err);
+    uint64_t expected_write_size = strict_strtoll(nargs[obj_name ? 2 : 3], 10, &err);
     if (!err.empty()) {
       cerr << "couldn't parse expected_write_size: " << err << std::endl;
       usage(cerr);
       return 1;
     }
-    ret = io_ctx.set_alloc_hint(oid, expected_object_size, expected_write_size);
+    ret = io_ctx.set_alloc_hint(*obj_name, expected_object_size, expected_write_size);
```
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/tools/rados/rados.cc:3396]: I've forgotten the usual check for emptiness of the `std::optional`. Implemented in following way for the sake of coherency with other commands:

```diff
@@ -3393,6 +3394,9 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
       usage(cerr);
       return 1;
     }
+    if (!obj_name) {
+      obj_name = nargs[1];
+    }
     ret = io_ctx.set_alloc_hint(*obj_name, expected_object_size, expected_write_size);
     if (ret < 0) {
       cerr << "error setting alloc-hint " << pool_name << "/" << prettify(*obj_name) << ": "
```

Thanks for pointing this out, Kefu!

File: src/tools/rados/rados.cc (hunk 1 above)
Comment (rzarzynski) [reply src/tools/rados/rados.cc:3385]: Stayed with the ternary operator to be consistent with other commands.

Comment (tchaikov) [reply src/tools/rados/rados.cc:3385]: i just don't like the magic numbers. difficult to reason about and maintain.

**Pull request merged**